inc/dashboard/namespace.php

```diff
@@ -0,0 +1,456 @@
+ *
+ * @param int $days How many days worth of analytics to fetch. Defaults to 7.
+ */
+function get_views_list( int $days = 7 ) : array {
+	$date_start = time() - ( $days * DAY_IN_SECONDS );
+
+	$query = [
+		'query' => [
+			'bool' => [
+				'filter' => [
+					[
+						// Query from the current site.
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						// We're interested in views and conversions.
+						'terms' => [
+							'event_type.keyword' => [
+								'experienceView',
+								'conversion',
+							],
+						],
+					],
+					[
+						// Limit to date range.
+						'range' => [
+							'event_timestamp' => [
+								'gte' => $date_start * 1000,
+							],
+						],
+					],
+				],
+			],
+		],
+		'aggs' => [
+			'blocks' => [
+				'terms' => [
+					// Get the block data. This will give us the key for the block, which is stored as the post slug.
+					'field' => 'attributes.clientId.keyword',
+					'size' => 10000, // Use arbitrary large size that is more than we're likely to need.
+				],
+				'aggs' => [
+					'views' => [
+						'filter' => [
+							'term' => [
+								'event_type.keyword' => 'experienceView',
+							],
+						],
+						'aggs' => [
+							'uniques' => [
+								'cardinality' => [
+									'field' => 'endpoint.Id.keyword',
+								],
+							],
+						],
+					],
+					'conversions' => [
+						'filter' => [
+							'term' => [
+								'event_type.keyword' => 'conversion',
+							],
+						],
+						'aggs' => [
+							'uniques' => [
+								'cardinality' => [
+									'field' => 'endpoint.Id.keyword',
+								],
+							],
+						],
+					],
+					'conversionRate' => [
+						'bucket_script' => [
+							'buckets_path' => [
+								'views' => 'views>uniques',
+								'conversions' => 'conversions>uniques',
+							],
+							'script' => 'params.conversions / params.views',
+						],
+					],
+				],
+			],
+		],
+		'size' => 0,
+		'sort' => [
+			'event_timestamp' => 'desc',
+		],
+	];
+
+	$key = sprintf( 'views:list:days:%d', $days );
+	$cache = wp_cache_get( $key, 'altis-xbs' );
+
+	if ( $cache ) {
+		return $cache;
+	}
+
+	$result = Utils\query( $query );
+
+	if ( ! $result ) {
+		$data = [];
+
+		wp_cache_set( $key, $data, 'altis-xbs', MINUTE_IN_SECONDS );
+		return $data;
+	}
+
+	$data = get_aggregate_data( $result['aggregations']['blocks']['buckets'] ?? [] );
+	wp_cache_set( $key, $data, 'altis-xbs', 5 * MINUTE_IN_SECONDS );
+
+	return $data;
+}
+
+/**
+ * Alter the default WP_Query to change the sort order.
+ *
+ * @param \WP_Query $query The WP_Query object.
+ *
+ * @return \WP_Query The updated WP_Query object.
+ */
+function modify_views_list_query( $query ) {
+	// Bail if we aren't looking at the XB Insights page or if we're ordering by block title.
+	if (
+		$query->get( 'post_type' ) !== Blocks\POST_TYPE ||
+		in_array( $query->get( 'orderby' ), [ 'title', 'author', 'last_modified' ], true )
```

Comment:
What if I sort by some post meta or ID or any other option that WP supports? When we have a known smaller set of values that we want to process in a specific way it makes more sense to return early based on those rather than a handful of other possible options.

I also find it a bit neater and less potentially confusing to have a bunch of separate `if` statements for returning early quite often. It's more verbose but the logic is easier to see and step through when using xdebug for example. Basically I just try and avoid `||` operators as much as possible!

Reply:
> What if I sort by some post meta or ID or any other option that WP supports? When we have a known smaller set of values that we want to process in a specific way it makes more sense to return early based on those rather than a handful of other possible options.

Not sure I follow the suggestion here.

Reply:
I got there in the end.... 😅